encoding.bs

```diff
@@ -1305,6 +1350,404 @@ must run these steps:
+byteReadable
+  .pipeThrough(decoder)
+  .pipeTo(textWritable);</code></pre>
+
+  <p>If the <a for=TextDecoderStream>error mode</a> is "<code>fatal</code>" and
+  <a for=TextDecoderStream>encoding</a>'s <a for=/>decoder</a> returns <a>error</a>, both
+  {{GenericTransformStream/readable}} and {{GenericTransformStream/writable}} will be errored with a
+  {{TypeError}}.
+</dl>
+
+<p>The
+<dfn constructor for=TextDecoderStream id=dom-textdecoderstream><code>TextDecoderStream(<var>label</var>,
+<var>options</var>)</code></dfn> constructor, when invoked, must run these steps:
+
+<ol>
+ <li><p>Let <var>encoding</var> be the result of <a>getting an encoding</a> from <var>label</var>.
+
+ <li><p>If <var>encoding</var> is failure or <a>replacement</a>, then <a>throw</a> a {{RangeError}}.
+
+ <li><p>Let <var>dec</var> be a new {{TextDecoderStream}} object.
+
+ <li><p>Set <var>dec</var>'s <a for=TextDecoderStream>encoding</a> to <var>encoding</var>.
+
+ <li><p>If <var>options</var>'s <code>fatal</code> member is true, then set <var>dec</var>'s
+ <a for=TextDecoderStream>error mode</a> to "<code>fatal</code>".
+
+ <li><p>If <var>options</var>'s <code>ignoreBOM</code> member is true, then set <var>dec</var>'s
+ <a for=TextDecoderStream>ignore BOM flag</a>.
+
+ <li>
+  <p>Set <var>dec</var>'s <a for=TextDecoderStream>decoder</a> to a new <a for=/>decoder</a>
+  for <var>dec</var>'s <a for=TextDecoderStream>encoding</a>, and set <var>dec</var>'s
+  <a for=TextDecoderStream>stream</a> to a new <a for=/>stream</a>.
+
+ <li><p>Let <var>startAlgorithm</var> be an algorithm that takes no arguments and returns nothing.
+
+ <li><p>Let <var>transformAlgorithm</var> be an algorithm which takes a <var>chunk</var> argument
+ and runs the <a>decode and enqueue a chunk</a> algorithm with <var>dec</var> and
+ <var>chunk</var>.
+
+ <li><p>Let <var>flushAlgorithm</var> be an algorithm which takes no arguments and runs the <a>flush
+ and enqueue</a> algorithm with <var>dec</var>.
+
+ <li><p>Let <var>transform</var> be the result of calling
+ <a abstract-op>CreateTransformStream</a>(<var>startAlgorithm</var>, <var>transformAlgorithm</var>,
+ <var>flushAlgorithm</var>).
+
+ <li><p>Set <var>dec</var>'s <a for=TextDecoderStream>transform</a> to <var>transform</var>.
+
+ <li><p>Return <var>dec</var>.
+</ol>
+
+<p>The <dfn id=concept-tds-decode-and-enqueue>decode and enqueue a chunk</dfn> algorithm, given a
+{{TextDecoderStream}} <var>dec</var> and a <var>chunk</var>, runs these steps:
+
+<ol>
+ <li><p>If <a>Type</a>(<var>chunk</var>) is not Object, or <var>chunk</var> does not have an
+ \[[ArrayBufferData]] internal slot, or <a abstract-op>IsDetachedBuffer</a>(<var>chunk</var>) is
+ true, or <a abstract-op>IsSharedArrayBuffer</a>(<var>chunk</var>) is true, then return a new
+ promise rejected with a {{TypeError}} exception.
+
+ <li><p><a>Push</a> a <a lt="get a copy of the buffer source">copy of</a> <var>chunk</var> to
+ <var>dec</var>'s <a for=TextDecoderStream>stream</a>.
```

Comment:
Hmm one day we should do a transfer variant, of this and all other web platform APIs...